cli/azd/pkg/project/framework_service_maven.go

```diff
@@ -153,6 +174,58 @@ func (m *mavenProject) Package(
 	}, nil
 }
 
+// funcAppDir returns the directory of the function app packaged by azure-functions-maven-plugin for the given service.
+//
+// The app is typically packaged under target/azure-functions.
+func (m *mavenProject) funcAppDir(ctx context.Context, svc *ServiceConfig) (string, error) {
+	svcPath := svc.Path()
+	// The staging directory for azure-functions-maven-plugin is target/azure-functions.
+	// It isn't configurable, but this may change in the future: https://github.com/microsoft/azure-maven-plugins/issues/1968
+	functionsStagingRel := filepath.Join("target", "azure-functions")
+	functionsStagingDir := filepath.Join(svcPath, functionsStagingRel)
+
+	// A conventional azure-functions-maven-plugin project will have the property 'functionAppName' in pom.xml,
+	// with its property value is passed to azure-functions-maven-plugin as 'appName'.
+	appName, err := m.mavenCli.GetProperty(ctx, "functionAppName", svcPath)
+	if err != nil && !errors.Is(err, maven.ErrPropertyNotFound) {
+		return "", err
+	}
+
+	if appName != "" {
+		funcDir := filepath.Join(functionsStagingDir, appName)
+		if _, err := os.Stat(funcDir); err == nil {
+			return funcDir, nil
+		} else if !errors.Is(err, os.ErrNotExist) {
+			return "", fmt.Errorf("checking for function app directory: %w", err)
+		}
+	}
+
+	entries, err := os.ReadDir(functionsStagingDir)
+	if err != nil {
+		return "", fmt.Errorf("reading azure-functions directory: %w", err)
+	}
```

Comment:
is this error ever possible after checking os.Stat() in prev lines for this path+appName? 
Should we first check if dir exists and then dir+appName?

Reply:
I think this reads fine to me.

On line 197, we check if `target/functions/<appName>` exists. If so, we would use that.
If the directory does not exist, as on line 204, we fallback to checking if `target/azure-functions` has exactly one child directory, if so, we would use that.

The checks are not redundant, errors are mutually possible in either check. The reason why `target/functions/<appName>` is only used when it exists, is that the source of the information,`functionAppName` is also  *heuristic* in nature. The fact that a maven project specifies this property, doesn't mean it is always configured that way. The actual source of truth is the configured `appName` passed to the `azure-functions-maven-plugin`, but that isn't obtainable by `azd`.